syntax = "proto3";
package ragllm_service;

option go_package = "github.com/milkyhoop/flow-executor/internal/proto/ragllm_service";

import "google/protobuf/empty.proto";

// ===== ALL MESSAGE DEFINITIONS FIRST =====

// Legacy messages (Keep for backward compatibility)
message Ragllm_serviceRequest {
  string user_id = 1;
  string input = 2;
}

message Ragllm_serviceResponse {
  string status = 1;
  string result = 2;
}

// Embedding messages
message EmbeddingRequest {
  string text = 1;
}

message EmbeddingResponse {
  repeated float embedding = 1;
}

// Search messages
message SearchRequest {
  string tenant_id = 1;
  repeated float embedding = 2;
  int32 top_k = 3;
}

message SearchResponse {
  repeated SearchResult documents = 1;
}

message SearchResult {
  string id = 1;
  string title = 2;
  float score = 3;
}

// Primary response generation messages
message GenerateAnswerRequest {
  string question = 1;
  string tenant_id = 2;
  string mode = 3;                    // "conversation", "customer_service", "execution", "synthesis"
  string faq_context = 4;             // CRITICAL: Formatted FAQ context from orchestrator
  string intelligence_level = 5;       // "synthesis" (Tier 2) or "deep" (Tier 3)
  string model = 6;                   // "gpt-3.5-turbo", "gpt-4", etc.
  string user_id = 7;                 // Optional: for customer service mode detection
}

message GenerateAnswerResponse {
  string answer = 1;
  string mode = 2;
  string model_used = 3;              // Actual model used for response
  float confidence = 4;               // Response confidence score
  int32 processing_time_ms = 5;       // Processing time in milliseconds
}

// Synthesis messages
message SynthesizeRequest {
  string query = 1;
  string tenant_id = 2;
  string faq_context = 3;             // Formatted FAQ context string
  string synthesis_type = 4;          // "friendly", "professional", "concise"
  string model = 5;                   // Optional: specific model to use
}

message SynthesizeResponse {
  string synthesized_text = 1;
  string model_used = 2;
  float confidence = 3;
  string synthesis_type = 4;          // Echo back the synthesis type used
  int32 processing_time_ms = 5;       // Processing time in milliseconds
}

// ===== SERVICE DEFINITION (AFTER ALL MESSAGES) =====
service RagLlmService {
  // Health and utility methods
  rpc HealthCheck (google.protobuf.Empty) returns (google.protobuf.Empty);
  rpc DoSomething (Ragllm_serviceRequest) returns (Ragllm_serviceResponse);
  
  // Embedding and search methods  
  rpc GenerateEmbedding (EmbeddingRequest) returns (EmbeddingResponse);
  rpc SearchSimilarDocuments (SearchRequest) returns (SearchResponse);
  
  // Primary response generation methods
  rpc GenerateAnswer (GenerateAnswerRequest) returns (GenerateAnswerResponse);
  rpc Synthesize (SynthesizeRequest) returns (SynthesizeResponse);
}